form/src/main/java/org/dhis2/form/bindings/RuleExtensions.kt

```diff
@@ -283,54 +286,103 @@ fun ProgramRuleVariable.toRuleVariable(
             RuleVariableCalculatedValue.create(
                 name() ?: "",
                 dataElement()?.uid() ?: trackedEntityAttribute()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         ProgramRuleVariableSourceType.TEI_ATTRIBUTE ->
             RuleVariableAttribute.create(
                 name() ?: "",
                 trackedEntityAttribute()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         ProgramRuleVariableSourceType.DATAELEMENT_NEWEST_EVENT_PROGRAM_STAGE ->
             RuleVariableNewestStageEvent.create(
                 name() ?: "",
                 dataElement()?.uid() ?: "",
                 programStage()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         ProgramRuleVariableSourceType.DATAELEMENT_NEWEST_EVENT_PROGRAM ->
             RuleVariableNewestEvent.create(
                 name() ?: "",
                 dataElement()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         ProgramRuleVariableSourceType.DATAELEMENT_CURRENT_EVENT ->
             RuleVariableCurrentEvent.create(
                 name() ?: "",
                 dataElement()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         ProgramRuleVariableSourceType.DATAELEMENT_PREVIOUS_EVENT ->
             RuleVariablePreviousEvent.create(
                 name() ?: "",
                 dataElement()?.uid() ?: "",
-                valueType
+                valueType,
+                useCodeForOptionSet,
+                options
             )
 
         else -> throw IllegalArgumentException("Unsupported variable ")
     }
 }
 
+fun getOptions(
+    useCodeForOptionSet: Boolean,
+    dataElementUid: String?,
+    trackedEntityAttributeUid: String?,
+    attributeRepository: TrackedEntityAttributeCollectionRepository,
+    dataElementRepository: DataElementCollectionRepository,
+    optionRepository: OptionCollectionRepository
+): List<Option> {
+    if (useCodeForOptionSet) {
+        return emptyList()
+    }
+
+    if (dataElementUid != null) {
+        return dataElementRepository.uid(dataElementUid).blockingGet()?.optionSet()?.uid()
+            ?.let { optionSetUid ->
+                optionRepository.byOptionSetUid().eq(optionSetUid).blockingGet()
+            }?.map { option -> Option(option.name(), option.code()) } ?: emptyList()
+    } else if (trackedEntityAttributeUid != null) {
+        attributeRepository.uid(trackedEntityAttributeUid).blockingGet()?.optionSet()?.uid()
```

Comment:
It seems there is a missing "return" here

Reply:
fixed!